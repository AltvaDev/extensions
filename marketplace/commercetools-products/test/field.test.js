'use strict';

const React = require('react');
const renderer = require('react-test-renderer');

const { mockComponent, flushPromises } = require('./helpers.js');

jest.mock('@contentful/forma-36-react-components', () => ({
  Button: mockComponent('button'),
  Spinner: mockComponent('spinner')
}));

jest.mock('react-sortable-hoc', () => ({
  SortableContainer: component => component
}));

jest.mock('../src/product-card.js', () => ({
  SortableProductCard: mockComponent('sortable-product-card')
}));

jest.mock('../src/apollo.js', () => ({
  getApolloClient: function() {
    class ApolloClient {}

    return Promise.resolve(new ApolloClient());
  }
}));

const { CommerceToolsField } = require('./../src/field.js');

describe('CommerceToolsField', function() {
  const parameters = {
    projectKey: 'a-project-key',
    clientId: '',
    clientSecret: '',
    apiUri: '',
    authUri: '',
    locale: 'en'
  };

  test('Renders a spinner while loading ApolloClient', function() {
    const extension = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer
      .create(<CommerceToolsField extension={extension} parameters={parameters} isSingle={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test.each([['single', true], ['multiple', false]])(
    'Renders empty state for %s product(s)',
    function(_, isSingle) {
      const extension = {
        field: {
          getValue() {
            return undefined;
          },
          onValueChanged() {},
          onIsDisabledChanged() {}
        }
      };

      const component = renderer.create(
        <CommerceToolsField extension={extension} parameters={parameters} isSingle={isSingle} />
      );

      return flushPromises().then(() => expect(component.toJSON()).toMatchSnapshot());
    }
  );

  test.each([
    ['single', true, 'ABCDE1234'],
    ['multiple', false, ['ABCDE1234', 'FGHHI1234', 'JKLMN5678']]
  ])('Renders with existing values for %s product(s)', function(_, isSingle, existingValue) {
    const extension = {
      field: {
        getValue() {
          return existingValue;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const component = renderer.create(
      <CommerceToolsField extension={extension} parameters={parameters} isSingle={isSingle} />
    );

    return flushPromises().then(() => expect(component.toJSON()).toMatchSnapshot());
  });

  describe('disabled state', function() {
    test.each([['single', true], ['multiple', false]])(
      'Renders empty state for %s product(s)',
      function(_, isSingle) {
        const extension = {
          field: {
            getValue() {
              return undefined;
            },
            onValueChanged() {},
            onIsDisabledChanged(fn) {
              fn(true);
            }
          }
        };

        const component = renderer.create(
          <CommerceToolsField extension={extension} parameters={parameters} isSingle={isSingle} />
        );

        return flushPromises().then(() => expect(component.toJSON()).toMatchSnapshot());
      }
    );

    test.each([
      ['single', true, 'ABCDE1234'],
      ['multiple', false, ['ABCDE1234', 'FGHHI1234', 'JKLMN5678']]
    ])('Renders with existing values for %s product(s)', function(_, isSingle, existingValue) {
      const extension = {
        field: {
          getValue() {
            return existingValue;
          },
          onValueChanged() {},
          onIsDisabledChanged(fn) {
            fn(true);
          }
        }
      };

      const component = renderer.create(
        <CommerceToolsField extension={extension} parameters={parameters} isSingle={isSingle} />
      );

      return flushPromises().then(() => expect(component.toJSON()).toMatchSnapshot());
    });
  });
});
