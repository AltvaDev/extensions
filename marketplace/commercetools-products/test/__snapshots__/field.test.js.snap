// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommerceToolsField Renders a spinner while loading ApolloClient 1`] = `
<div
  className="spinner"
  size="large"
/>
`;

exports[`CommerceToolsField Renders empty state for multiple product(s) 1`] = `
Array [
  <div />,
  <div
    buttonType="muted"
    className="button"
    disabled={false}
    icon="ShoppingCart"
    onClick={[Function]}
    size="small"
  >
    Select Products
  </div>,
]
`;

exports[`CommerceToolsField Renders empty state for single product(s) 1`] = `
Array [
  <div />,
  <div
    buttonType="muted"
    className="button"
    disabled={false}
    icon="ShoppingCart"
    onClick={[Function]}
    size="small"
  >
    Select Product
  </div>,
]
`;

exports[`CommerceToolsField Renders with existing values for multiple product(s) 1`] = `
Array [
  <div>
    <div
      className="sortable-product-card"
      index={0}
      itemIndex={0}
      locale="en"
      removeItem={[Function]}
      sku="ABCDE1234"
      sortable={true}
    />
    <div
      className="sortable-product-card"
      index={1}
      itemIndex={1}
      locale="en"
      removeItem={[Function]}
      sku="FGHHI1234"
      sortable={true}
    />
    <div
      className="sortable-product-card"
      index={2}
      itemIndex={2}
      locale="en"
      removeItem={[Function]}
      sku="JKLMN5678"
      sortable={true}
    />
  </div>,
  <div
    buttonType="muted"
    className="button"
    disabled={false}
    icon="ShoppingCart"
    onClick={[Function]}
    size="small"
  >
    Select Products
  </div>,
]
`;

exports[`CommerceToolsField Renders with existing values for single product(s) 1`] = `
<div>
  <div
    className="sortable-product-card"
    index={0}
    itemIndex={0}
    locale="en"
    removeItem={[Function]}
    sku="ABCDE1234"
    sortable={false}
  />
</div>
`;
